libmscore/changeMap.cpp

```diff
@@ -0,0 +1,547 @@
+                        event.cachedStartVal = DEFAULT_VALUE;
+                        }
+                  }
+
+            // Phase 2: cache an end value for the ramp
+            // If there's no set velocity change:
+            if (event.value == 0) {
+                  auto nextChangeEventIter = i;
+                  nextChangeEventIter++;
+                  // There's a chance that the next event is a fix at the same tick as the
+                  // start of the current ramp. If so, get the next event, which is assured
+                  // to be a different (larger) tick
+                  if (nextChangeEventIter != end() && nextChangeEventIter.key() == tick)
+                        nextChangeEventIter++;
+
+                  // If this is the last event, there is no change
+                  if (nextChangeEventIter == end()) {
+                        event.cachedEndVal = event.cachedStartVal;
+                        }
+                  else {
+                        // Search for a fixed event at the next event point
+                        bool foundFix = false;
+                        for (auto& nextChangeEvent : values(nextChangeEventIter.key())) {
+                              if (nextChangeEvent.type == ChangeEventType::FIX) {
+                                    event.cachedEndVal = nextChangeEvent.value;
+                                    foundFix = true;
+                                    break;
+                                    }
+                              }
+
+                        // We haven't found a fix, so there must be a ramp. What does the user want?
+                        // A good guess would to be to interpolate, but that might get complex, so just ignore
+                        // this ramp and set the ending to be the same as the start.
+                        // TODO: implementing some form of smart interpolation would be nice.
+                        if (!foundFix) {
+                              event.cachedEndVal = event.cachedStartVal;
+                              }
+                        }
+                  }
+            else {
+                  event.cachedEndVal = event.cachedStartVal + event.value;
+                  }
+
+            // And finally... if something's wrong, make it not wrong
+            if ((event.cachedStartVal > event.cachedEndVal && event.direction == ChangeDirection::INCREASING) ||
+                (event.cachedStartVal < event.cachedEndVal && event.direction == ChangeDirection::DECREASING)) {
+                  event.cachedEndVal = event.cachedStartVal;
+                  }
+            }
+
+      }
+
+//---------------------------------------------------------
+//   cleanup
+//---------------------------------------------------------
+
+void ChangeMap::cleanup()
+      {
+      if (cleanedUp)
+            return;
+
+      // qDebug() << "Before cleanup:";
+      // dump();
+
+      cleanupStage0();
+      cleanupStage1();
+      cleanupStage3();
```

Comment:
For `cleanupStage2()` naming, in its current state this sequence makes me think "but where is 2?" :)